model/lifecycle.go

```diff
@@ -57,25 +57,85 @@ type VersionToRestart struct {
 // SetVersionActivation updates the "active" state of all builds and tasks associated with a
 // version to the given setting. It also updates the task cache for all builds affected.
 func SetVersionActivation(versionId string, active bool, caller string) error {
-	builds, err := build.Find(
-		build.ByVersion(versionId).WithFields(build.IdKey),
-	)
+	version, err := VersionFindOneId(versionId)
 	if err != nil {
-		return errors.Wrapf(err, "getting builds for version '%s'", versionId)
+		return errors.Wrapf(err, "getting version '%s'", versionId)
 	}
-	buildIDs := make([]string, 0, len(builds))
-	for _, build := range builds {
-		buildIDs = append(buildIDs, build.Id)
+	if version == nil {
+		return errors.Wrapf(err, "version '%s' not found", versionId)
 	}
-
-	// Update activation for all builds before updating their tasks so the version won't spend
-	// time in an intermediate state where only some builds are updated
-	if err = build.UpdateActivation(buildIDs, active, caller); err != nil {
-		return errors.Wrapf(err, "setting activation for builds in version '%s'", versionId)
+	if err = version.SetActivated(active); err != nil {
+		return errors.Wrapf(err, "setting activated for version '%s'", versionId)
+	}
+	changed, err := setTaskActivationForVersion(versionId, active, false, nil, caller)
+	if err != nil {
+		return errors.Wrapf(err, "setting activation for tasks in version '%s'", versionId)
+	}
+	if active && changed && evergreen.IsFinishedVersionStatus(version.Status) {
+		if err = version.UpdateStatus(evergreen.VersionCreated); err != nil {
+			return errors.Wrapf(err, "setting status for version '%s'", versionId)
+		}
 	}
+	return nil
+}
 
-	return errors.Wrapf(setTaskActivationForBuilds(buildIDs, active, false, nil, caller),
-		"setting activation for tasks in version '%s'", versionId)
+func setTaskActivationForVersion(versionId string, active, withDependencies bool, ignoreTasks []string, caller string) (bool, error) {
+	// If activating a task, set the ActivatedBy field to be the caller
+	if active {
+		q := bson.M{
+			task.VersionKey: versionId,
+			task.StatusKey:  evergreen.TaskUndispatched,
+		}
+		if len(ignoreTasks) > 0 {
+			q[task.IdKey] = bson.M{"$nin": ignoreTasks}
+		}
+		tasksToActivate, err := task.FindAll(db.Query(q).WithFields(task.IdKey, task.DependsOnKey, task.ExecutionKey, task.BuildIdKey))
+		if err != nil {
+			return false, errors.Wrap(err, "getting tasks to activate")
+		}
+		if withDependencies {
+			dependOn, err := task.GetRecursiveDependenciesUp(tasksToActivate, nil)
+			if err != nil {
+				return false, errors.Wrap(err, "getting recursive dependencies")
+			}
+			tasksToActivate = append(tasksToActivate, dependOn...)
+		}
+		if len(tasksToActivate) > 0 {
```

Comment:
I'm wondering if in the !withDependencies case we could not query tasks at all, but just update tasks/builds that match that query.

We could refactor ActivateTasks to take in a query, and for other usages  of ActivateTasks (that _do_ use a task list) make some helper function like `byTaskIds(taskList)` that returns `IdKey: bson.M{"$in": taskIDs}` if that makes sense